### PR TITLE
Allow reloading variable qty of ammo

### LINKED ISSUE #15122: Misleading "out of ammo" due to no magazine
Got a drop of 5x50mm and SPIW for it, both are shown as green, but reloading results in "out of ammo".
I assume this is because I lack the magazine.

### PR DESCRIPTION
**Pull request by mugling**
- `KEY_LEFT` and `KEY_RIGHT` allow the `amount` of ammo to reload to be changed
- Ammo belts can be assembled, extended and dissassembled
- Some reload failure messages are improved (including fixes #15122)
- Reloading a magazine already contained within a gun has the appropriate time penalty
- Provide option to part reload tools/guns

Further work (separate PR):
- Prompt when multiple types of ammo belt could be assembled (not just default to first)
- Handle interruption of reloading (load a proportional quantity of ammo rather than fail outright)
- Implement magazine adaptors for converting tools to belt feed

### CODE REVIEW

**Comment by mugling** [comment]
Works sufficient for merge testing.

Marked `CR` as I want to build upon this idea (potentially as a separate PR) to include support for ammo belts and also displaying concise item descriptions below the menu via `uimenu::desc_enabled`. Suggestions/comments regarding this?

**Comment by mugling** [comment]
**REBASED**

**Comment by Coolthulhu** [comment]
It is possible to specify 0 amount

**Comment by Coolthulhu** [comment]
Flamethrower reloads as if it had RELOAD_ONE.

**Comment by Coolthulhu** [comment]
The "craft ammo belt" really doesn't belong with other item functions and should be an iuse of linkage instead.

**Comment by mugling** [comment]
> Flamethrower reloads as if it had RELOAD_ONE.

Can you expand upon this?

> The "craft ammo belt" really doesn't belong with other item functions and should be an iuse of linkage instead.

Sounds good

**Comment by Coolthulhu** [comment]
> Can you expand upon this?

* Spawn flamethrower and wield it
* Spawn contained gasoline
* Reload flamethrower with "reload wielded item"
* Takes a long time, loads a single charge for every reload attempt

**Comment by mugling** [comment]
All relevant fixes now committed so presuming no further issues is ready to merge

**Comment by mugling** [comment]
Anything else to do?

File: src/player.cpp [hunk 1]
```diff
@@ -9831,6 +9832,37 @@ bool player::can_use( const item& it, bool interactive ) const {
...
+    }
+
+    if( it.magazine_integral() ) {
+        if( !ammo.empty() ) {
+            if( it.ammo_data() ) {
+                if( it.ammo_data()->id != ammo ) {
+                    return false;
+                }
+            } else {
+                auto at = item_controller->find_template( ammo );
```
Comment (BevapDin) [inline src/player.cpp]: You used `item::find_type` in the new iuse_actor, why not here? That way you don't need to include item_factory.h

File: src/game.cpp [hunk 2]
```diff
@@ -11361,7 +11365,16 @@ void game::reload( int pos )
             if( it->ammo_remaining() > 0 && it->ammo_remaining() == it->ammo_capacity() ) {
                 add_msg( m_info, _( "The %s is already fully loaded!" ), it->tname().c_str() );
                 return;
-            } // intentional fall-through
+            }
+            if( it->is_ammo_belt() ) {
+                auto linkage = it->type->magazine->linkage;
+                if( linkage != "NULL" && !g->u.has_charges( linkage, 1 ) ) {
+                    add_msg( m_info, _( "You need at least one %s to reload the %s!" ),
+                                     item::find_type( linkage )->nname( 1 ).c_str(), it->tname().c_str() );
```
Comment (BevapDin) [inline src/game.cpp]: You can use `item::nname(linkage, 1)` instead of `item::find_type( linkage )->nname( 1 )`

File: src/player.cpp (hunk 1 above)
Comment (mugling) [inline src/player.cpp]: Good point

**Comment by Coolthulhu** [comment]
Causes debugmsgs with generic guns.

**Comment by mugling** [comment]
@Coolthulhu give me a few mins to look into that

**Comment by mugling** [comment]
@Coolthulhu fixed in 147c0a6

**Comment by Coolthulhu** [comment]
Why is Jenkins complaining about Makefile?

**Comment by mugling** [comment]
I just merged #16168

**Pull request merged**